rllib/agents/ppo/ppo.py

```diff
@@ -462,3 +477,54 @@ def execution_plan(
         return StandardMetricsReporting(train_op, workers, config).for_each(
             lambda result: warn_about_bad_reward_scales(config, result)
         )
+
+    @ExperimentalAPI
+    def training_iteration(self) -> ResultDict:
+        # Collect SampleBatches from sample workers
+        rollouts = synchronous_parallel_sample(self.workers)
+
+        # Concatenate the SampleBatches from each worker into one large SampleBatch
+        rollouts = SampleBatch.concat_samples(rollouts)
+        self._counters[NUM_AGENT_STEPS_SAMPLED] += rollouts.agent_steps()
+        self._counters[NUM_ENV_STEPS_SAMPLED] += rollouts.env_steps()
+        # Standardize advantages
+        rollouts = standardize_fields(rollouts, ["advantages"])
+        # Train
+        if self.config["simple_optimizer"]:
+            train_results = train_one_step(self, rollouts)
+        else:
+            train_results = multi_gpu_train_one_step(self, rollouts)
+
+        # For each policy: update KL scale and warn about possible issues
+        for i, (policy_id, policy_info) in enumerate(train_results.items()):
+            # Update KL loss with dynamic scaling
+            # for each (possibly multiagent) policy we are training
+            kl_divergence = policy_info[LEARNER_STATS_KEY].get("kl")
+            self.get_policy(policy_id).update_kl(kl_divergence)
+
+            # Warn about excessively high value function loss
+            scaled_vf_loss = (
+                self.config["vf_loss_coeff"] * policy_info[LEARNER_STATS_KEY]["vf_loss"]
+            )
+            policy_loss = policy_info[LEARNER_STATS_KEY]["policy_loss"]
+            if (
+                self.config.get("model", {}).get("vf_share_layers")
+                and scaled_vf_loss > 100
+            ):
+                logger.warning(
+                    "The magnitude of your value function loss for policy: {} is "
+                    "extremely large ({}) compared to the policy loss ({}). This "
+                    "can prevent the policy from learning. Consider scaling down "
+                    "the VF loss by reducing vf_loss_coeff, or disabling "
+                    "vf_share_layers.".format(policy_id, scaled_vf_loss, policy_loss)
+                )
+            # Warn about bad clipping configs
+            mean_reward = rollouts["rewards"][rollouts["agent_index"] == i].mean()
```

Comment:
Great idea just doing this here and in this fashion! True, we don't really need entire episodes for this avg-reward estimate.

Reply:
should we log 1 in N times or something? won't this flood the training logs?

Reply:
I think @smorad fixed this by a `log_once` if-block, which seems like always the best option for these warnings.